sllm/serve/schedulers/storage_aware_scheduler.py

```diff
@@ -154,3 +206,294 @@ async def _control_loop(self):
+                            # Copy the previous plan without including the current instance
+                            dp[i][j] = copy.deepcopy(dp[i - 1][j])
+                    else:
+                        logger.info(
+                            f"Instance {i} cannot fit into GPU capacity {j}"
+                        )
+                        # If the instance cannot fit, carry forward the previous plan
+                        dp[i][j] = copy.deepcopy(dp[i - 1][j])
+
+            # Return the optimal Migration plans
+            minLatency = float("inf")
+            optimal_plans = None
+            for plans in dp[numInstances]:
+                logger.info(f"Total latency: {plans.total_latency}")
+                if (
+                    plans.total_latency < minLatency
+                    and plans.evictedGPUs >= gpu_shortage
+                ):
+                    minLatency = plans.total_latency
+                    optimal_plans = plans
+            logger.info(f"Migration plans: {optimal_plans}")
+            return optimal_plans
+        except Exception as e:
+            logger.error(f"Failed to get migration plans: {e}")
+            return None
+
+    async def mark_resource(
+        self, model_name: str, instance_id: str, node_id: int
+    ) -> bool:
+        logger.info(f"Model {model_name} instance {instance_id} marked")
+        async with self.metadata_lock:
+            if model_name not in self.model_instance:
+                self.model_instance[model_name] = {}
+            self.model_instance[model_name][instance_id] = node_id
+        return node_id
+
+    async def set_model_scheduler_config(
+        self, model_name: str, scheduler_config: Mapping
+    ) -> bool:
+        logger.info(f"Setting scheduler config for model {model_name}")
+        async with self.metadata_lock:
+            self.model_scheduler_config[model_name] = scheduler_config
+        return True
+
+    def _get_model_loading_time(
+        self,
+        model_name: str,
+        model_size: int,
+        hardware_info: Mapping,
+        node_waiting_time: float,
+        pinned_memory_pool: Mapping,
+    ) -> float:
+        latency = 0
+        if model_name not in pinned_memory_pool:
+            latency += (
+                node_waiting_time + model_size / hardware_info["disk_bandwidth"]
+            )
+            logger.info(
+                f"Loading model {model_name} will take {latency} seconds"
+            )
+        else:
+            latency += model_size / hardware_info["pcie_bandwidth"]
+            logger.info(
+                f"Loading model {model_name} will take {latency} seconds"
+            )
+        return latency
+
+    def _get_migration_target(
+        self,
+        model_name: str,
+        num_gpu: int,
+        source_node_id: int,
+        worker_nodes: Mapping,
+        model_info: Mapping,
+        store_info: Mapping,
+        hardware_info: Mapping,
+    ) -> Optional[MigrationPlan]:
```

Comment:
This type hint seems not correct, return type is `(node_id, loading_time)`.

Reply:
Fixed in the latest commit.